### PR TITLE
Datamon context

### PR DESCRIPTION
**Pull request by kerneltime**
This will be a big change and starting the review process early
Todo

- [x] Complete the IO paths usage of bundles

- [x] Move internal documentation to GitHub

- [x] Update tests.

```
1. Introduce the Context: Named bucket sets that uniquely identify a
datamon instance.
2. Introduce ContextStores: storage.Store interface implementation set
for a Context.
3. Introduce Configuration for CLI. A remote repository of contexts.
4. Change CLI to only depend on the configuration bucket to detect a context
and use it. 
```

I removed Kubernetes CSI plugin as it is no longer in use and do not want to maintain the code. 
Signed-off-by: Ritesh H Shukla <ritesh@oneconcern.com>

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -26,12 +26,7 @@ linters:
     - gochecknoinits
     - gochecknoglobals
     #- golint
-    - godox
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply .golangci.yml]: I was having a problem where linting was failing locally as the corresponding binaries were not found.. Will try it again.

File: cmd/datamon/cmd/bundle.go [hunk 2]
```diff
@@ -22,44 +23,45 @@ var bundleCmd = &cobra.Command{
...
 
-func setLatestOrLabelledBundle(ctx context.Context, store storage.Store) error {
+func setLatestOrLabelledBundle(ctx context.Context, remote context2.DatamonContextStores) error {
 	switch {
-	case params.bundle.ID != "" && params.label.Name != "":
-		return fmt.Errorf("--" + addBundleFlag(nil) + " and --" + addLabelNameFlag(nil) + " flags are mutually exclusive")
-	case params.bundle.ID == "" && params.label.Name == "":
-		key, err := core.GetLatestBundle(params.repo.RepoName, store)
+	case datamonFlags.bundle.ID != "" && datamonFlags.label.Name != "":
+		return fmt.Errorf("--" + addBundleFlag(nil) + " and --" + addLabelNameFlag(nil) + " datamonFlags are mutually exclusive")
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply cmd/datamon/cmd/bundle.go]: Will do.

File: cmd/datamon/cmd/config.go [hunk 3]
```diff
@@ -1,39 +1,70 @@
...
+	if err != nil {
+		wrapFatalln("failed to unmarshal", err)
+	}
+	flags.context.Descriptor = contextDescriptor
+}
+
+func (c *CLIConfig) createContext(datamonContext model.Context, flags *flagsT) {
+	configStore, err := gcs.New(context.Background(), flags.core.Config, config.Credential)
+	if err != nil {
+		wrapFatalln("failed to create config store. ", err)
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply cmd/datamon/cmd/config.go]: thanks

File: cmd/datamon/cmd/config.go [hunk 4]
```diff
@@ -1,39 +1,70 @@
...
 	}
-	if params.repo.BlobBucket == "" {
-		params.repo.BlobBucket = config.Blob
-		if params.repo.BlobBucket == "" {
-			wrapFatalln("blob bucket not set in config or as a cli param", nil)
-		}
+	err = context2.CreateContext(configStore, datamonContext)
+	if err != nil {
+		wrapFatalln("failed to store context. ", err)
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply cmd/datamon/cmd/config.go]: thanks

File: cmd/datamon/cmd/config_generate.go [hunk 5]
```diff
@@ -11,33 +11,35 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const datamonDir = ".datamon2"
+
 var configGen = &cobra.Command{
 	Use:   "create",
 	Short: "Create a config",
-	Long:  "Create a config to use for datamon. Config file will be placed in $HOME/.datamon/datamon.yaml",
+	Long:  "Create a config to use for datamon. CLIConfig file will be placed in $HOME/" + datamonDir + "/datamon.yaml",
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply cmd/datamon/cmd/config_generate.go]: thanks

File: pkg/core/bundle.go [hunk 6]
```diff
@@ -33,9 +35,8 @@ var MemProfDir string
 type Bundle struct {
 	RepoID                      string
 	BundleID                    string
-	MetaStore                   storage.Store
 	ConsumableStore             storage.Store
-	BlobStore                   storage.Store
+	contextStores               context2.DatamonContextStores
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle.go]: Only things in model package should be marshaled and un marshaled.

File: pkg/core/bundle.go [hunk 7]
```diff
@@ -116,21 +117,47 @@ func Repo(r string) BundleOption {
...
 	return func(b *Bundle) {
-		b.BlobStore = store
+		b.contextStores = cs
 	}
 }
+func (b *Bundle) BlobStore() storage.Store {
+	return getBlobStore(b.contextStores)
+}
+
+func getBlobStore(stores context2.DatamonContextStores) storage.Store {
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle.go]: will rename.. my Java past leaked a bit.. I do want getter and setters vs direct access to variables.

File: pkg/core/bundle_list.go [hunk 8]
```diff
@@ -417,12 +419,12 @@ func getBundleAsync(repo string, store storage.Store,
...
 // GetLatestBundle returns the latest bundle descriptor from a repo
-func GetLatestBundle(repo string, store storage.Store) (string, error) {
-	e := RepoExists(repo, store)
+func GetLatestBundle(repo string, stores context2.DatamonContextStores) (string, error) {
+	e := RepoExists(repo, stores)
 	if e != nil {
 		return "", e
 	}
-	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxMetaFilesToProcess)
+	ks, _, err := getMetaStore(stores).KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxMetaFilesToProcess)
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle_list.go]: The contract is that all members of contextStores should be available. Will add checks in CLI and see if other checks make sense. This is an on or off kind of bug.

File: pkg/context/context.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+type DatamonContextStores struct {
+	wal       storage.Store
+	readLog   storage.Store
+	blob      storage.Store
+	metadata  storage.Store
+	vMetadata storage.Store
+	_         struct{}
+}
+
+func NewDatamonContextStores(wal storage.Store, readLog storage.Store, blob storage.Store, metadata storage.Store, vMetadata storage.Store) *DatamonContextStores {
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/context/context.go]: Makes sense.
The only place where we optionally set stores is in tests, in real world usage all stores will be configured and they should be New. Will see how I can enforce this further.

File: pkg/context/context.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+package context
+
+import (
+	"github.com/oneconcern/datamon/pkg/model"
+	"github.com/oneconcern/datamon/pkg/storage"
+)
+
+// ContextStores for datamon.
+type DatamonContextStores struct {
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/context/context.go]: Will do.

File: pkg/context/context.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/context/context.go]: For cases where the tests only need to set one, they use the setter.

File: pkg/core/bundle.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle.go]: Yes this is an internal attribute of a bundle that should be public and only decided at the time of creation.

**Comment by ransomw1c** [comment]
### scurrilous vandalism in progress

* contextualized parameters |-> stores

File: cmd/datamon/cmd/bundle_download.go [hunk 11]
```diff
@@ -22,42 +22,41 @@ var BundleDownloadCmd = &cobra.Command{
...
+		bundleOpts = append(bundleOpts, core.ConcurrentFilelistDownloads(
+			datamonFlags.bundle.ConcurrencyFactor/filelistDownloadsByConcurrencyFactor))
+
+		bundle := core.NewBundle(core.NewBDescriptor(),
+			bundleOpts...,
 		)
 
-		if params.bundle.NameFilter != "" {
-			nameFilterRe, err = regexp.Compile(params.bundle.NameFilter)
+		var nameFilterRe *regexp.Regexp
```
Comment (ransomw1c) [inline cmd/datamon/cmd/bundle_download.go]: ##### nitpick

i think the ["TypeSpec"](https://golang.org/ref/spec#Declarations_and_scope) of `if`is `if` itself -- contrast with `for .. range ..`, possibly? -- such that the declaration, either in Hindley-Milner or not, within the `{}` or on its own line has (mild) implications

Comment (kerneltime) [reply cmd/datamon/cmd/bundle_download.go]: Yes the variable declaration can be moved into the loop, it was initially defined at the very top of the func.

**Comment by kerneltime** [comment]
> ### scurrilous vandalism in progress
> * contextualized parameters |-> remote stores

Please refer to the docs added. The selection of stores for a context should never be configurable.

File: pkg/core/bundle_list.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle_list.go]: https://github.com/oneconcern/datamon/pull/280/files#diff-9924a8d557b824af4026ae9144b9140cR255 will fail if any one of the stores is not available.

File: pkg/core/bundle.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerneltime) [reply pkg/core/bundle.go]: I wanted to have one place where the selection of which bucket to be used is done.. I probably over did it.

File: .golangci.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kerneltime) [reply .golangci.yml]: Fixed

**Comment by kerneltime** [comment]
Moving to https://github.com/oneconcern/datamon/pull/282

**Pull request closed without merging**